### DISCUSSION TITLE
Shortcuts for different dictionary groups under GNOME/Wayland (Linux)

### DISCUSSION THREAD

**Discussion by saligari** [goldendict/goldendict#1646] [Q&A] [answered]
Hello,

I've been using Fedora w/ GNOME & Wayland for the past couple of years. Under this system goldendict's shortcuts for popups etc don't seem to be working, so I've been using a custom keybinding:

> sh -c 'goldendict "$(wl-paste)"'

to force the popup with a search of my clipboard (copied word).

I was wondering if there was a way that I could have different keybindings to call the popup to search within different dictionary groups.

Thanks!

**Comment by vedgy** [comment, marked as answer]
Create several shortcuts for different groups like this:
```bash
sh -c 'goldendict --popup-group-name="name of group" "$(wl-paste)"'
```
See also GoldenDict help (F1) => 14 Command line switches.

**Reply by saligari** [reply]
Thanks, doesn't work very well, neither with --popup-group-name nor with --group-name. Like, most of the time it doesn't work and searches within the group that was previously open.

**Reply by vedgy** [reply]
Works every time on my Manjaro GNU/Linux, X11, Xfce system. Is the group set to the value of the `popup-group-name` argument in the Group combobox at the top of the popup window? If yes, this could be some race condition on your system.

Messages are sent via *qt-single-application* API: first `setPopupGroup`, then `translateWord`: https://github.com/goldendict/goldendict/blob/0252e7bff671462cf4243e38251b1808b6b465c1/main.cc#L381
When the `setPopupGroup` message is received, the group is set synchronously: https://github.com/goldendict/goldendict/blob/0252e7bff671462cf4243e38251b1808b6b465c1/mainwindow.cc#L4022

Perhaps *qt-single-application* delivers messages out of order on your system. You could check by printing debug output when a message is received. Insert
1) `gdWarning( "translateWord: %s received", message.mid( 15 ).toUtf8().constData()  );` in the `if( message.left( 15 ) == "translateWord: " )` block
and
2) `gdWarning( "setPopupGroup: %s received", message.mid( 15 ).toUtf8().constData()  );` in the `if( message.left( 15 ) == "setPopupGroup: " )` block
in *mainwindow.cc*.

The output after running e.g. `goldendict --popup-group-name=Local test` in terminal should always be:
```
setPopupGroup: Local received
translateWord: test received
```

**Reply by nonwill** [reply]
> Perhaps _qt-single-application_ delivers messages out of order

If that's the issue, how about delay some milliseconds before sending the message when **_wasMessage_** is **_true_**? 
**_main.cc:_**
```
  if( !showHelpAndExit && app.isRunning() )
  {
    ....
    if( gdcl.needTranslateWord() )
    {
      if( wasMessage )
        app.processEvents( QEventLoop::AllEvents, 100 );
      else
        wasMessage = true;

     app.sendMessage( QString( "translateWord: " ) + gdcl.wordToTranslate() );
    }
```

**Reply by vedgy** [reply]
> If that's the issue, how about delay some milliseconds before sending the message when wasMessage is true?

An artificial delay is not a great solution. Since GoldenDict bundles *qt-single-application*, it can be patched to preserve the order of the messages. Another alternative is to send a single composite message instead of multiple small ones. But first let's wait for debug output feedback. Maybe the bug lies elsewhere.

**Reply by nonwill** [reply]
> Maybe the bug lies elsewhere.

Yes, it is possible, I have not encountered the reported issue on Windows either.
